### PR TITLE
WIP: Restart config what

### PR DESCRIPTION
**Pull request by joakim-hove**
Started to build a configuration object of which keywords should be included in restart. This is work in progress and not ready for merge.

### CODE REVIEW

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp [hunk 1]
```diff
@@ -20,6 +20,8 @@
 #ifndef OPM_RESTART_CONFIG_HPP
 #define OPM_RESTART_CONFIG_HPP
 
+#include <vector>
+
 #include <boost/date_time/gregorian/gregorian_types.hpp>
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Seems to me like this boost include can go away too. Happy days!

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 2]
```diff
@@ -195,6 +205,118 @@ namespace Opm {
...
 
+    void RestartConfig::update( size_t step, const RestartSchedule& rs) {
+        if (step == 0)
+            this->restart_schedule->updateInitial( rs );
+        else
+            this->restart_schedule->update( step, rs );
+    }
+
+
+    const std::vector<std::string> RestartConfig::integerKeywords = {""     ,      //  1 BASIC=?? - ignored
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: Why is this a member in the first place?

Unless there's a very good reason for it, this should probably be `static constexpr const char* integerKeywords[] = { ... }.` (or anon namespace as you see fit).

Why? Well, first of all it's a simpler data type for what is static (as in binary region) in the binary anyway, and you use indexing for lookup by design. Second, it opens up a few optimisations for no syntactic cost.

Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: > Why is this a member in the first place?

Old habits.

> static constexpr const char* integerKeywords[] = { ... }. (or anon namespace as you see fit).

I'll update - guess I'll go for the anon namespace.

This patch has some miles to go anyway.

As a side comment: I would encourage anyone skeptical to the value of tests to refactor this code; I have been guided semicolon by semicolon by the tests!

**Comment by joakim-hove** [comment]
OK -this should be conent-wise ready; did not have time for rebasing.

**Comment by jokva** [comment]
I'll have a look.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #include <boost/lexical_cast.hpp>
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: Good opportunity to replace lexical_cast, maybe?

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 4]
```diff
@@ -50,10 +51,97 @@ namespace Opm {
...
+                                                    "SURFBLK",    // 26
+                                                    "PCOW,PCOG",  // 27 - Multiple, split on "," [Special cased]
+                                                    "STREAM",     // 28 STREAM=?? - not supported
+                                                    "RK",         // 29
+                                                    "VELOCITY",   // 30
+                                                    "COMPRESS" }; // 31
+
+
+        static void insertKeyword(std::set<std::string>& keywords, const std::string& keyword) {
+            const auto pos = keyword.find( "=" );
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: There's actually a `basic_string::find( CharT, size_type pos = 0 )` overload. Since we're looking for a character we can use that rather than string-based. Minor, but should be slightly faster.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 5]
```diff
@@ -70,52 +158,65 @@ namespace Opm {
...
-        // is done to make flow write restart file at the same intervals
-        // as eclipse for the Norne data set. There might be some rules
-        // we are missing here.
-        if( 0 == basic ) return {};
+                rs = RestartSchedule( step, basic, freq );
+            }
 
-        if( items.size() > FREQ_index ) // if frequency is set
-            freq = std::stoul( items[ FREQ_index ] );
+            updateKeywords( step , fun::map( []( const std::string& s) { return std::stoi(s); } , items ));
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: `fun::map( std::stoi, items )`.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp [hunk 6]
```diff
@@ -20,7 +20,165 @@
...
+
+     RESTART=1 : As RPTRST,BASIC=1
+     RESTART>1 : As RPTRST,BASIC=2
+     RESTART>2 : Flow is added to restart file
+     RESTART>3 : Fluid in place is added to restart file
+     RESTART=6 : Restart file for every timestep.
+
+
+
+  RPTSOL: The RPTSOL keyword is very similar to the RPTCHED keyword,
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Typo? I presume you meana RPTSCHED

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp [hunk 7]
```diff
@@ -114,7 +273,36 @@ namespace Opm {
...
+        RestartSchedule() = default;
+        RestartSchedule( size_t sched_restart);
+        RestartSchedule( size_t step, size_t b, size_t freq);
+        bool writeRestartFile( size_t timestep , const TimeMap& timemap) const;
+        bool operator!=(const RestartSchedule& rhs) const;
+        bool operator==( const RestartSchedule& rhs ) const;
+
+
+
+        //private:
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Re-privatise the members?

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp [hunk 8]
```diff
@@ -155,55 +347,15 @@ namespace Opm {
...
-
-        std::shared_ptr<DynamicState<restart>> m_restart_output_config;
+        void handleScheduleSection( const SCHEDULESection& schedule);
+        void update( size_t step, const RestartSchedule& rs);
+        void addKeywords( size_t step, const std::vector<std::string>& keywords);
+        void updateKeywords( size_t step, const std::vector<int>& integer_controls);
+        void handleRPTRST( const DeckKeyword&, size_t );
+        static RestartSchedule rptsched( const DeckKeyword& );
+
+        std::shared_ptr<DynamicState<RestartSchedule>> restart_schedule;
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Can we kill the `shared_ptr`s here?

**Comment by jokva** [comment]
I guess we should re-enable the old tests as well, no?

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 9]
```diff
@@ -50,10 +51,97 @@ namespace Opm {
                 && std::all_of( x.begin() + 1, x.end(), is_digit );
         }
 
+
+        std::set<std::string> notSupported = {"NORST", "SFREQ" , "STREAM"};
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: Just a general observation here: I don't think this needs to be a set. It's three elements, and it doesn't change at all. We can make it a constexpr array and just use `any_of`, meaning the algorithm for unsupported later becomes:

```
bool unsupported( const std::string& kw ) {
    static constexpr const char* keywords[] = { "NORST", "SFREQ", "STREAM" };   
return std::any_of( std::begin( keywords ), std::end( keywords ), std::equal< std::string >() );
}
[...]

if( !unsupported( keyword ) ) insertKeyword( keyword );
```

Obviously, making it an array or something doesn't change the real benefit here which is providing a function that determines if a keyword is unsupported or not, that can also be accomplished with a string. It does lift the burden of comparison from the insert algorithm.

**Comment by jokva** [comment]
Personally I feel this gets a bit unwieldy now. Little bit too much going on in too many branches.

It was already pretty hard to follow, but the flow was largely linear in the sense that there were no multiple branches, only exit conditions.

If I get the overall idea of the patch right, we want to
1. Register at what timesteps to restart (this was already supported)
2. Register *what* to output at some time step.

which is user configurable through the same two keywords. Assuming this is correct, I want to ask the following:

1. How many keywords control this behaviour? The 26 `integerKeywords` or more? Is it feasible to not use a string set to represent this?
2. Can we split the operations and assembly up a bit? In its current state it feels a bit ad-hoc and case-A-then-maybe-B-unless-C-then-D.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 10]
```diff
@@ -155,19 +256,86 @@ namespace Opm {
...
+      keywords - there are some of them.
+    */
+
+    void RestartConfig::addKeywords( size_t step, const std::vector<std::string>& keywords) {
+        if (keywords.size() > 0) {
+            std::set<std::string> kw_set = this->restart_keywords->back();
+            for (const auto& kw : keywords )
+                insertKeyword( kw_set , kw );
+
+            if (step == 0)
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: This can be encapsulated in `update`.

**Comment by jokva** [comment]
Furthermore I'm not quite sure I see the value of the add/update keyword interface. Care to shed some light on it?

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 11]
```diff
@@ -155,19 +256,86 @@ namespace Opm {
...
+      to remove keywords with the string based menonics.
+    */
+
+    void RestartConfig::updateKeywords( size_t step, const std::vector<int>& integer_controls) {
+        std::set<std::string> keywords = this->restart_keywords->back();
+
+        for (size_t index = 0; index < integer_controls.size(); index++) {
+            if (index == 26) {
+                if (integer_controls[index] > 0) {
+                    insertKeyword( keywords , "PCOW" );
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: This is confusing. If POCW+PCOG hasn't been added because integer_controls[26] (can be done outside of the loop by the way, clearer that way) <= 0, why would you need to erase them?

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 12]
```diff
@@ -184,35 +352,65 @@ namespace Opm {
...
+                                             );
 
-            restart_config.update( current_step, rs );
+                update( current_step , rs );
+            }
         }
+    }
+
+    bool RestartSchedule::writeRestartFile( size_t input_timestep , const TimeMap& timemap) const {
+        if (this->rptsched_restart_set && (this->rptsched_restart > 0))
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: Nice.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp (hunk 5 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: I thought so to - did not work at first attempt; will try again.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp (hunk 6 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Yes.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp (hunk 7 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp]: Yes - that was the plan; forgot it.

File: opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp (hunk 10 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: >This can be encapsulated in update.

You mean the `DynamicState::update()`?

Comment (jokva) [inline opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp]: No, `RestartConfig::update`

**Comment by joakim-hove** [comment]
OK @jokva - it's all yours!

Observe that there are two failing tests - these test test that the `RESTART=n` mnenonic on the `RPTSOL` and `RPTSCHED` keywords can configure that the `FIP` and `FLOW` keywords should be added to list of keywords. That is only implemented in the test - in true TDD style.

**Comment by jokva** [comment]
Great. Since the tests are failing I'll use this branch as a starting point rather than merge into master and then go ham, and my final PR will be the one schedule for merge.

**Comment by joakim-hove** [comment]
> and my final PR will be the one schedule for merge.

Yes - that was my plan; do merge your own future PR when you are happy with it.

**Comment by jokva** [comment]
Will do, great work!

**Comment by pgdr** [comment]
Am I to understand that we can close this PR and that #887 is the one that eventually will be merged?

**Comment by jokva** [comment]
Yes.
